node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+	async fn run<Context>(mut ctx: Context) -> SubsystemResult<()>
+	where
+		Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+	{
+		// startup: register the network protocol with the bridge.
+		ctx.send_message(AllMessages::NetworkBridge(
+			NetworkBridgeMessage::RegisterEventProducer(Self::PROTOCOL_ID, network_update_message),
+		))
+		.await?;
+
+		// work: process incoming messages from the overseer and process accordingly.
+		let mut tracker = Tracker::default();
+		loop {
+			{
+				let message = ctx.recv().await?;
+				match message {
+					FromOverseer::Communication { msg } => {
+						// another subsystem created this signed availability bitfield messages
+						match msg {
+							// relay_message a bitfield via gossip to other validators
+							BitfieldDistributionMessage::DistributeBitfield(
+								hash,
+								signed_availability,
+							) => {
+								trace!(target: "bitd", "Processing DistributeBitfield");
+								let job_data = &mut tracker.per_relay_parent.get_mut(&hash).expect("Overseer does not send work items related to relay parents that are not part of our workset. qed");
+
+								let validator = {
+									job_data
+										.validator_set
+										.get(signed_availability.validator_index() as usize)
+										.expect("Our own validation index exists. qed")
+								}
+								.clone();
+
+								let peer_views = &mut tracker.peer_views;
+								let msg = BitfieldGossipMessage {
+									relay_parent: hash,
+									signed_availability,
+								};
+
+								relay_message(&mut ctx, job_data, peer_views, validator, msg)
+									.await?;
+							}
+							BitfieldDistributionMessage::NetworkBridgeUpdate(event) => {
+								trace!(target: "bitd", "Processing NetworkMessage");
+								// a network message was received
+								if let Err(e) =
+									handle_network_msg(&mut ctx, &mut tracker, event).await
+								{
+									warn!(target: "bitd", "Failed to handle incomming network messages: {:?}", e);
+								}
+							}
+						}
+					}
+					FromOverseer::Signal(OverseerSignal::StartWork(relay_parent)) => {
+						trace!(target: "bitd", "Start {:?}", relay_parent);
+						// query basic system parameters once
+						// @todo assumption: these cannot change within a session
+						let (validator_set, signing_context) =
+							query_basics(&mut ctx, relay_parent).await?;
+
+						let _ = tracker.per_relay_parent.insert(
+							relay_parent,
+							PerRelayParentData {
+								signing_context,
+								validator_set,
+								..Default::default()
+							},
+						);
+					}
+					FromOverseer::Signal(OverseerSignal::StopWork(relay_parent)) => {
+						trace!(target: "bitd", "Stop {:?}", relay_parent);
+						// @todo assumption: it is good enough to prevent additional work from being
+						// scheduled, the individual futures are supposedly completed quickly
+						let _ = tracker.per_relay_parent.remove(&relay_parent);
```

Comment:
It might be better to do nothing here and avoid a race condition.

What happens if we `StopWork(H)` here, but our peers still think `H` is in our view? The `NetworkBridgeUpdate::OurViewChange` is generated based on `StartWork` and `StopWork` messages and is only sent to the other subsystems once we've dispatched a notification to our peers of our view changing. There is still the potential for a race, but it's small. 

So it would be better to just clean up based on the view change notification - what was in our view that isn't anymore?

Reply:
But to eliminate the issue there should be a grace period, during which we do not punish incoming messages for which we were previously interested. I am not sure if waiting for having dispatched a message makes much of a difference, since the other node would need to send a response instantaneously which is not possible (even if I assume that dispatch here means we received a TCP SYN ACK rather than just having written the message to a tcp buffer layer).